src/folding/hypernova/circuit.rs

```diff
@@ -42,15 +43,60 @@ impl<F: PrimeField> SumMulsGammaPowsEqSigmaGadget<F> {
     }
 }
 
+/// Gadget to compute $\Sigma_{i=1}^{q} c_i * \Pi_{j \in S_i} theta_j$.
+pub struct SumCiMulProdThetajGadget<F: PrimeField> {
+    _f: PhantomData<F>,
+}
+
+impl<F: PrimeField> SumCiMulProdThetajGadget<F> {
+    /// Computes $\Sigma_{i=1}^{q} c_i * \Pi_{j \in S_i} theta_j$
+    ///
+    /// # Arguments
+    /// - `c_i`: vector of $c_i$ values
+    /// - `thetas`: vector of pre-processed $\thetas[j]$ values corresponding to a particular `ccs.S[i]`
+    ///
+    /// # Notes
+    /// This is a part of the second term of the sum that $\mathcal{V}$ has to compute at section 5, step 5 of "A multi-folding scheme for CCS".
+    /// The first term is computed by `SumMulsGammaPowsEqSigmaGadget::sum_muls_gamma_pows_eq_sigma`.
+    /// This is a doct product between a vector of c_i values and a vector of pre-processed $\theta_j$ values, where $j$ is a value from $S_i$.
+    /// Hence, this requires some pre-processing of the $\theta_j$ values, before running this gadget.
+    pub fn sum_ci_mul_prod_thetaj(c_i: Vec<FpVar<F>>, thetas: Vec<Vec<FpVar<F>>>) -> FpVar<F> {
+        let mut result = FpVar::<F>::zero();
+        for (i, c_i) in c_i.iter().enumerate() {
+            let prod = &thetas[i].iter().fold(FpVar::one(), |acc, e| acc * e);
+            result += c_i * prod;
+        }
+        result
+    }
+}
+
+/// Returns a vector of thetas for a corresponding $S_i$
+/// An helper function to run before computing $\Pi_{j \in S_i} \theta_j$ in a circuit.
+pub fn get_prepared_thetas<C: CurveGroup>(
```

Comment:
Currently this method is only used in the tests, I understand that in a future iteration it will be used by some non-test method, right?
(if not, I would suggest to move it inside the tests)

Additionally, I'm wondering that since the method just does a loop & push, maybe it could be easier to read the code without jumping between methods and directly doing the loop computation inline (something like a oneline iterator doing the push without requiring a whole new method). How do you see it?

Reply:
Yup I agree, updated with [ac92a8b](https://github.com/privacy-scaling-explorations/folding-schemes/pull/56/commits/ac92a8b38aae2a58964000c77805286b635f8a38)